documentation/smg/16_serialiser.md

```diff
@@ -0,0 +1,229 @@
+- This will be followed by the data serialised in a relevant form.
+- There is no footer required as the header contains all relevant information as to the size of the serialised object.
+- Tags are `uint8` data and the "magic" numbers are described in detail in the section [Tag Format][tag] below.
+- Dimensions are serialised as `uint32`
+
+The single exception to this header is that of `function_handles` which are purely scalar and the rank component of the tag is used to specify the type of function handle:
+
+Handle type         |   tag   | value
+--------------------|---------|-------
+Simple/Class simple | 32 + 25 | 57
+Anonymous           | 64 + 25 | 89
+Scoped              | 96 + 25 | 121
+
+## Tag format
+
+### Tag Structure
+
+The tag in the new format is a 1-byte (`uint8`) tag where the top 3 bits give the rank of the data, the bottom 5 give the type-tag of the data
+
+| 128 | 64 | 32 | 16 | 8 | 4 | 2 | 1 |
+|:---:|:--:|:--:|:--:|:-:|:-:|:-:|:-:|
+|  R  | R  | R  | T  | T | T | T | T |
+
+- R - Rank
+- T - Type
+
+### Type Tags
+Tag| Meaning
+---|----------------------
+  0| LOGICAL
+  1| CHAR
+  2| MATLAB_STRING
+  3| DOUBLE
+  4| SINGLE
+  5| INT8
+  6| UINT8
+  7| INT16
+  8| UINT16
+  9| INT32
+ 10| UINT32
+ 11| INT64
+ 12| UINT64
+ 13| COMPLEX_DOUBLE
+ 14| COMPLEX_SINGLE
+ 15| COMPLEX_INT8
+ 16| COMPLEX_UINT8
+ 17| COMPLEX_INT16
+ 18| COMPLEX_UINT16
+ 19| COMPLEX_INT32
+ 20| COMPLEX_UINT32
+ 21| COMPLEX_INT64
+ 22| COMPLEX_UINT64
+ 23| CELL
+ 24| STRUCT
+ 25| FUNCTION_HANDLE
+ 26| VALUE_OBJECT
+ 27| HANDLE_OBJECT_REF
+ 28| ENUM
+ 29| SPARSE_LOGICAL
+ 30| SPARSE_DOUBLE
+ 31| SPARSE_COMPLEX_DOUBLE
+
+### Serialisation formats in detail
+
+#### Simple data
+Simple data (type tags 0-12) are serialised as:
+
+_Rank_ |     tag    | dim1 | dim2 | ... | Data
+-------|------------|------|------|:---:|------
+_NULL_ | 32 + type  | 0
+_0_    |    type    |      |      |     | Data
+_1_    | 32 + type  | nElem|      |     | Data
+_2_    | 64 + type  | dim1 | dim2 |     | Data
+_N_    | N<<5 + type| dim1 | dim2 | ... | Data
+
+#### Complex Data
+Complex numerical data (type tags 13-22) are serialised as:
+
+_Rank_ |     tag    | dim1 | dim2 | ... |    Data   |   ...
+-------|------------|------|------|:---:|-----------|----------
+_NULL_ | 32 + type  | 0    |      |     |           |
+_0_    |    type    |      |      |     | Real Data | Imag Data
+_1_    | 32 + type  | nElem|      |     | Real Data | Imag Data
+_2_    | 64 + type  | dim1 | dim2 |     | Real Data | Imag Data
+_N_    | N<<5 + type| dim1 | dim2 | ... | Real Data | Imag Data
+
+#### Sparse data
+Sparse numeric data (type tags 29-31) are serialised as:
+
+_Rank_ |   Type  |     tag  | dim1 | dim2 | Data | ... |    ...    | ...
+-------|---------|----------|------|------|:----:|:---:|-----------|------
+_2_    | Real    | 64 + type| dim1 | dim2 |   i  |  j  | Data      |
+_2_    | Bool    | 64 + type| dim1 | dim2 |   i  |  j  | Data      |
```

Comment:
You said above that you were excluding the data description from the header format. In this section you have been continuing to include the data - which may be a good thing for clarification purposes - but at this point you extend that data to include the sparseness indices and this reader was then wondering if i and j are part of the header or part of the data - and it may be ambiguous which it should be. My instinct is that your tendency to include the data may be an inclination to document the data structure - would it be easier to include that in this document? Apart from that suggestion, I think you need to be consistent in either including or excluding the data, and at least saying where its structure is documented.

Reply:
This is the documentation for the serialisation of the data, not the header. The point of the header section is that it is a standardised form as we can see throughout here. If anything the header should be removed here, but I think it provides a more complete view of what you're going to see in the serialised data, rather than splitting them.

Reply:
@oerc0122 The whole document may be for the serialisation of the data, but we are in the section (I think, following the ##etc structure) for the tag format, which is part of the header format. In fact most of the document appears to be header and tag. You may want to think about the section titles and any introductory text to the sections. I would personally put tag as a sub-head of header as that is where tags are used. Happy to discuss. But a section starting ## Data would be useful to indicate when you move onto that from the header.